lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+            if (std.c.posix_memalign(&aligned_ptr, alignment, size) != 0) {
+                @panic("posix_memalign");
+            }
+            errdefer std.c.free(aligned_ptr);
+
+            if (control.default_value) |value| {
+                // default value: copy the content to newly allocated object.
+                @memcpy(@ptrCast([*]u8, aligned_ptr), @ptrCast([*]u8, value), size);
+            } else {
+                // no default: return zeroed memory.
+                @memset(@ptrCast([*]u8, aligned_ptr), 0, size);
+            }
+
+            self.slots[index] = @ptrCast(*c_void, aligned_ptr);
+        }
+
+        return self.slots[index].?;
+    }
+};
+
+// global stucture for ThreadStorage.
+// it provides thread-safety for on-demand storage of Thread Objects.
+const ThreadStorage = struct {
+    var key: std.c.pthread_key_t = undefined;
+    var init_once = std.once(ThreadStorage.init);
+
+    /// Return a per thread ObjectArray with at least the expected index.
+    pub inline fn get_array(allocator: *Allocator, index: usize) !*ObjectArray {
+        if (ThreadStorage.getspecific()) |array| {
+            // ensure the array is big enough for the wanted index
+            try array.ensureSize(index);
+            return array;
+        } else {
+            // a new array should contains at least 16 objects.
+            const size = std.math.max(16, index);
+
+            // create a new array and store it.
+            var array: *ObjectArray = try ObjectArray.initAlloc(allocator, size);
+            ThreadStorage.setspecific(array);
+            return array;
+        }
+    }
+
+    /// Return casted thread specific value.
+    inline fn getspecific() ?*ObjectArray {
+        return @ptrCast(
+            ?*ObjectArray,
+            @alignCast(
+                @alignOf(?*ObjectArray),
+                std.c.pthread_getspecific(ThreadStorage.key),
+            ),
+        );
+    }
+
+    /// Set casted thread specific value.
+    inline fn setspecific(new: ?*ObjectArray) void {
+        if (std.c.pthread_setspecific(ThreadStorage.key, @ptrCast(*c_void, new)) != 0)
+            @panic("pthread_setspecific");
+    }
+
+    /// Initialize pthread_key_t.
+    fn init() void {
+        if (std.c.pthread_key_create(&ThreadStorage.key, ThreadStorage.deinit) != 0)
+            @panic("pthread_key_create");
+    }
+
+    /// Invoked by pthread specific destructor. the passed argument is the ObjectArray pointer.
+    fn deinit(arrayPtr: *c_void) callconv(.C) void {
+        var array = @ptrCast(
+            *ObjectArray,
+            @alignCast(@alignOf(*ObjectArray), arrayPtr),
+        );
+        array.deinit();
+    }
+};
+
+// global counter for keeping track of requested indexes.
+const Index = struct {
+    var next: usize = 1;
+    var mutex: std.c.pthread_mutex_t = std.c.PTHREAD_MUTEX_INITIALIZER;
+
+    /// Return a next free index, incrementing global counter.
+    pub fn new() usize {
```

Comment:
If you just need to pull a fresh index I'd turn this into a single LOCK+GET+INCREMENT+UNLOCK sequence.

Reply:
I am not sure to understand the exact context of the remark.

if it is just about the `new()` method, it is already done under locking (i will rename it to `new_locked()` to be more explicit).

if it is about the global index managment (`getIndex()`), grabing a full mutex would be too inefficient: the function is called at every threadvar access.

Reply:
> if it is just about the new() method, it is already done under locking (i will rename it to new_locked() to be more explicit).

Instead of requiring the caller to lock, call `new` then unlock the mutex my suggestion was to replace it with a single method like this:

```zig
fn next_index() usize {
    mutex.lock();
    defer mutex.unlock();
    const old = index;
    index += 1;
    return index;
}
```

After all you only need the mutex to make sure each index is unique.

Reply:
the general usage of the index is: "get the index but if 0 initialize it with the next global value". with the fact that several threads could race to read/set the index.

the simple method could be:
- lock globally all threads for the global counter (it will also serialize access on control index)
- check the control index
- if 0, get the next index (and incr the global value)
- return the index (and unlock)

it is would involve taking a lock for every access of the index. lock couldn't be taken per variable, as `emutls_control` doesn't have structure to hold it.

the approch taken by LLVM emutls.c (and so by me too), is to use atomic reading lockless (so with possible races). in the general case (index already initialized) there is no problem and access is quick. for first access, it needs locking (and race checking as two threads could have read `0` index value and want to initialize it). so it can't be a all-in-one function for getting the next index.

eventually I could merge `Index.new()` inside `control.get_index()` function.

Reply:
Nevermind then, if multiple threads can race to initialize the same `emutls_control` this makes sense.
Packing `mutex` with the index is misleading, one is tempted to assume its only task is to protect the global counter while here it serves as a synchronization point for every thread/TLS object.

Reply:
I reorganized the code. the global variable for the index has been put toplevel. and the mutex is a global variable inside `emutls_control` and `lcok()` and `unlock()` helpers.

and now I wonder if I should put the global variable `global_index` inside `emutls_control` too.